pkg/dns/aws/health.go

```diff
@@ -0,0 +1,252 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1alpha1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/route53/route53iface"
+	"github.com/rs/xid"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	idTag = "kuadrant.dev/healthcheck"
+
+	defaultHealthCheckPath             = "/"
+	defaultHealthCheckPort             = 80
+	defaultHealthCheckFailureThreshold = 3
+)
+
+var (
+	callerReference func(id string) *string
+)
+
+type Route53HealthCheckReconciler struct {
+	client route53iface.Route53API
+}
+
+var _ dns.HealthCheckReconciler = &Route53HealthCheckReconciler{}
+
+func NewRoute53HealthCheckReconciler(client route53iface.Route53API) *Route53HealthCheckReconciler {
+	return &Route53HealthCheckReconciler{
+		client: client,
+	}
+}
+
+func (r *Route53HealthCheckReconciler) Reconcile(ctx context.Context, spec dns.HealthCheckSpec, endpoint *v1alpha1.Endpoint) (dns.HealthCheckResult, error) {
+	healthCheck, exists, err := r.findHealthCheck(ctx, endpoint)
+	if err != nil {
+		return dns.HealthCheckResult{}, err
+	}
+
+	defer func() {
+		if healthCheck != nil {
+			endpoint.SetProviderSpecific(ProviderSpecificHealthCheckID, *healthCheck.Id)
+		}
+	}()
+
+	if exists {
+		status, err := r.updateHealthCheck(ctx, spec, endpoint, healthCheck)
+		if err != nil {
+			return dns.HealthCheckResult{}, err
+		}
+
+		return dns.NewHealthCheckResult(status, ""), nil
+	}
+
+	healthCheck, err = r.createHealthCheck(ctx, spec, endpoint)
+	return dns.NewHealthCheckResult(dns.HealthCheckCreated, fmt.Sprintf("Created health check with ID %s", *healthCheck.Id)), err
+}
+
+func (r *Route53HealthCheckReconciler) Delete(ctx context.Context, endpoint *v1alpha1.Endpoint) (dns.HealthCheckResult, error) {
+	healthCheck, found, err := r.findHealthCheck(ctx, endpoint)
+	if err != nil {
+		return dns.HealthCheckResult{}, err
+	}
+	if !found {
+		return dns.NewHealthCheckResult(dns.HealthCheckNoop, ""), nil
+	}
+
+	_, err = r.client.DeleteHealthCheckWithContext(ctx, &route53.DeleteHealthCheckInput{
+		HealthCheckId: healthCheck.Id,
+	})
+
+	if err != nil {
+		return dns.HealthCheckResult{}, err
+	}
+
+	endpoint.DeleteProviderSpecific(ProviderSpecificHealthCheckID)
+	return dns.NewHealthCheckResult(dns.HealthCheckDeleted, ""), nil
+}
+
+func (c *Route53HealthCheckReconciler) findHealthCheck(ctx context.Context, endpoint *v1alpha1.Endpoint) (*route53.HealthCheck, bool, error) {
+	id, hasId := getHealthCheckId(endpoint)
+	if !hasId {
+		return nil, false, nil
+	}
+
+	response, err := c.client.GetHealthCheckWithContext(ctx, &route53.GetHealthCheckInput{
```

Comment:
Could we cache this?

Reply:
maybe cache it with the generation of the policy that caused it? So it invalidates when the policy is changed?